File: app/models/service_settings/geocoding_schema.rb
# frozen_string_literal: true

module ServiceSettings
  class GeocodingSchema
    KOMOOT_HOST = Geocoding::Providers::KOMOOT_HOST
    CHIBIGEO_HOST = Geocoding::Providers::CHIBIGEO_HOST
    CHIBIGEO_BARE_HOST = Geocoding::Providers::CHIBIGEO_BARE_HOST
    HOST_FORMAT = %r{\A[a-z0-9_][a-z0-9._-]*(:\d+)?(/[a-z0-9._/-]*)?\z}

    def initialize(setting)
      @setting = setting
    end

    def normalize
      normalize_host
      force_https_only_hosts
      normalize_rps
    end

    def validate
      validate_provider
      validate_host
      validate_api_key
      validate_chibigeo_key
    end

    def komoot?
      Geocoding::Providers.komoot?(setting.provider, setting.config['host'])
    end

    def chibigeo?
      Geocoding::Providers.chibigeo?(setting.provider, setting.config['host'])
    end

    def paid?
      Geocoding::Providers.api_key_required?(setting.provider)
    end

    private

    attr_reader :setting

    def normalize_host
      host = setting.config['host']
      return if host.blank?

      normalized = host.to_s.strip.downcase.sub(%r{\Ahttps?://}, '').sub(%r{/+\z}, '')
      setting.config['host'] = normalized
    end

    # Runs after normalize_host: the rate ceiling is decided by the host, so it
    # has to be read in its canonical form.
    def normalize_rps
      rate = Geocoding::RateLimits.for(setting.provider, setting.config['host']).normalize(setting.config['rps'])

      if rate.nil?
        setting.config.delete('rps')
      else
        setting.config['rps'] = rate
      end
    end

    def force_https_only_hosts
      setting.config['use_https'] = true if PHOTON_HTTPS_ONLY_HOSTS.include?(bare_host)
    end

    def bare_host
      Geocoding::Providers.bare_host(setting.config['host'])
    end

    def validate_provider
      return if Geocoding::Providers::CHAIN.include?(setting.provider)

      setting.errors.add(:provider, :inclusion)
    end

    def validate_host
      host = setting.config['host']

      if Geocoding::Providers.host_required?(setting.provider) && host.blank?
        setting.errors.add(:base, :host_required, provider: Geocoding::Providers.name(setting.provider))
      elsif host.present? && !host.match?(HOST_FORMAT)
        setting.errors.add(:base, :host_invalid)
      end
    end

    def validate_api_key
      return unless Geocoding::Providers.api_key_required?(setting.provider)
      return if setting.api_key.present?

      setting.errors.add(:base, :api_key_required, provider: Geocoding::Providers.name(setting.provider))
    end

    def validate_chibigeo_key
      return unless chibigeo?
      return if setting.api_key.present?

      setting.errors.add(:base, :chibigeo_api_key_required)
    end
  end
end

File: app-phoenix/lib/dawarich/release_migrations/effects/support/geocoding_schema.ex
defmodule Dawarich.ReleaseMigrations.Effects.Support.GeocodingSchema do
  @moduledoc false

  alias Dawarich.ReleaseMigrations.Effects.Support.{Ruby, ServiceSetting}

  @host_required ~w[photon nominatim]
  @api_key_required ~w[geoapify locationiq]
  @names %{
    "photon" => "Photon",
    "geoapify" => "Geoapify",
    "nominatim" => "Nominatim",
    "locationiq" => "LocationIQ"
  }
  @komoot_host "photon.komoot.io"
  @chibigeo_bare_host "app.chibigeo.com"
  @https_only_hosts ~w[photon.dawarich.app photon.komoot.io app.chibigeo.com]
  @host_format ~r/\A[a-z0-9_][a-z0-9._-]*(:\d+)?(\/[a-z0-9._\/-]*)?\z/
  @messages %{
    host_required:
      "%{provider} needs a host. Enter your instance's hostname, for example photon.example.com.",
    host_invalid:
      "The host must be a bare hostname with an optional port or path — no http:// prefix, no spaces.",
    api_key_required: "%{provider} needs an API key.",
    chibigeo_api_key_required:
      "ChibiGeo needs an API key. The free Hobby plan includes 2,500 lookups per day — get your key at chibigeo.com."
  }

  def https_only_hosts, do: @https_only_hosts
  def messages, do: @messages

  def normalize(setting, key) do
    config = setting.config
    host = Ruby.index(config, "host")

    unless is_map(config),
      do:
        raise(
          Ruby.Unreproducible,
          "cannot reproduce Ruby's normalization of #{Ruby.instance(config)}"
        )

    config = if Ruby.blank?(host), do: config, else: Map.put(config, "host", normalize_host(host))
    host = Ruby.index(config, "host")

    config =
      if bare_host(host) in @https_only_hosts,
        do: Map.put(config, "use_https", true),
        else: config

    setting = %{setting | config: config}

    setting =
      if komoot?(setting.provider, host),
        do: ServiceSetting.drop_api_key(setting, key),
        else: setting

    case rate(setting.provider, host, Ruby.index(config, "rps")) do
      nil -> %{setting | config: Map.delete(config, "rps")}
      rps -> %{setting | config: Map.put(config, "rps", rps)}
    end
  end

  def validate(setting, key) do
    host = Ruby.index(setting.config, "host")
    name = @names[setting.provider]

    host_errors =
      cond do
        setting.provider in @host_required and Ruby.blank?(host) ->
          [String.replace(@messages.host_required, "%{provider}", name)]

        Ruby.present?(host) and not (is_binary(host) and host =~ @host_format) ->
          [@messages.host_invalid]

        true ->
          []
      end

    key_errors =
      if setting.provider in @api_key_required and
           Ruby.blank?(ServiceSetting.api_key(setting, key)),
         do: [String.replace(@messages.api_key_required, "%{provider}", name)],
         else: []

    chibigeo_errors =
      if chibigeo?(setting.provider, host) and Ruby.blank?(ServiceSetting.api_key(setting, key)),
        do: [@messages.chibigeo_api_key_required],
        else: []

    host_errors ++ key_errors ++ chibigeo_errors
  end

  def validate!(setting, key) do
    case validate(setting, key) do
      [] -> :ok
      errors -> raise Ruby.Error, "Validation failed: " <> Enum.join(errors, ", ")
    end
  end

  defp normalize_host(host) do
    host
    |> Ruby.to_s()
    |> Ruby.strip()
    |> String.downcase()
    |> String.replace(~r/\Ahttps?:\/\//, "")
    |> String.replace(~r/\/+\z/, "")
  end

  defp bare_host(host) when is_binary(host),
    do: host |> String.split("/") |> hd() |> String.split(":") |> hd()

  defp bare_host(_host), do: nil

  defp komoot?(provider, host), do: provider == "photon" and bare_host(host) == @komoot_host

  defp chibigeo?(provider, host),
    do: provider == "photon" and bare_host(host) == @chibigeo_bare_host

  defp rate(provider, host, value) do
    cond do
      komoot?(provider, host) -> 1.0
      chibigeo?(provider, host) -> clamp(number(value), 1.0, 1.0, 25.0)
      true -> clamp(number(value), nil, 0.1, 1000.0)
    end
  end

  defp number(value) when is_number(value), do: value / 1
  defp number(value) when is_binary(value), do: Ruby.float(value)
  defp number(_value), do: nil

  defp clamp(number, default, _min, _max) when number in [nil, :neg_infinity], do: default
  defp clamp(:infinity, _default, _min, max), do: max
  defp clamp(number, default, _min, _max) when number <= 0, do: default
  defp clamp(number, _default, min, max), do: number |> max(min) |> min(max)
end
